module_auto_update/models/module.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+import logging
+
+from odoo import api, fields, models
+from odoo.modules.module import get_module_path
+
+_logger = logging.getLogger(__name__)
+try:
+    from checksumdir import dirhash
+except ImportError:
+    _logger.debug('Cannot `import checksumdir`.')
+
+
+class Module(models.Model):
+    _inherit = 'ir.module.module'
+    _rec_name = 'name'
+
+    checksum_installed = fields.Char(
+        string="Installed Module Checksum",
+    )
+    checksum_dir = fields.Char(
+        string="Current Module Checksum",
+        compute='_compute_checksum_dir',
+    )
+    upgrade_available = fields.Boolean(
+        default=False,
+    )
+
+    @api.multi
+    def _compute_checksum_dir(self):
+        for r in self:
+            r.checksum_dir = dirhash(get_module_path(r.name), 'sha1')
+
+    @api.multi
+    def button_immediate_upgrade(self):
+        res = super(Module, self).button_immediate_upgrade()
+
+        self.post_upgrade_actions()
+
+        return res
+
+    @api.model
+    def check_upgrade_availability(self):
+        self.update_list()
+        self.determine_upgradability()
+
+    @api.depends('checksum_dir')
+    def determine_upgradability(self):
+        upgradable_modules = self.search(
+            [('state', 'in', ['installed', 'to upgrade', 'to remove'])]
+        )
+
+        for r in upgradable_modules:
+            if r.checksum_dir != r.checksum_installed:
+                r.upgrade_available = True
+
+    @api.depends('checksum_dir')
+    def post_upgrade_actions(self):
+        for r in self:
+            r.checksum_installed = r.checksum_dir
```

Comment:
@Yajo It's also called from `upgrade_module` in `base.module.upgrade`; `button_immediate_upgrade` isn't used for applying scheduled upgrades.